### DISCUSSION TITLE
Unit testing

### DISCUSSION THREAD

**Discussion by j-maas** [telegraf/telegraf#224] [Q&A]
I've recently discovered Telegraf and find it really interesting. However, I'm struggling to setup an environment in which I can test my bot.

Specificly, I don't understand how to generate a mock of the framework that I can use to test my bot. I would like the possibility of doing something along the lines of
```
describe('myBot', () => {
  let telegraf  
  beforeEach(() => {
    telegraf = new TelegrafMock(mybot)  
  }
  it('should echo hi', (done) => {
    telegraf.insertUpdate(Object.assign({text: 'hi'}, baseMessage)
      .then((response) => expect(response.text).toEqual('hi))
      .then(done)
  }
}
```

What do you suggest I can do to test my code?

**Comment by bostrot** [comment]
i dont really understand your question. when you want to test your code take a vps or raspberry pi install the telegraf framework and jus test it. if this is what you mean?

**Comment by j-maas** [comment]
No, what I would like to do is _not_ connect the bot to the Telegram server but still test its logic. This is because using a local replacement for the servers would make it much faster and "safer". Also I could not find any documentation on how to write unit tests for Telegraf, except the tests for the framework itself, but those didn't really help me unfortunately.

So, in summary, I am looking for a way to write (integration) tests and mock away the Telegram server.

**Comment by bostrot** [comment]
You looking for something like "cycle-telegram" (github)?

**Comment by j-maas** [comment]
I do not understand how [cycle-telegram](https://github.com/goodmind/cycle-telegram) relates to this. I simply want to know how to write unit and integration tests with the Telegraf framework.

Sorry, if I'm not clear enough. Maybe you can point me to what I can explain more?

**Comment by dotcypress** [comment]
@Y0hy0h we plan to build something like that in [this repo](https://github.com/telegraf/telegram-test-server)

**Comment by j-maas** [comment]
Awesome! Thank you!

**Comment by bostrot** [comment]
@Y0hy0h initially i meant something like https://github.com/jehy/telegram-test-api (messed the copy/paste thing up).

**Comment by j-maas** [comment]
@bostrot, this looks perfect! Thanks for the link!

**Comment by yookoala** [comment]
Wrote a mock telegraf context for unit test. I guess someone might find it useful:

```node
const Context = require('telegraf/core/context');
const Telegram = require('telegraf/core/network/client');

/**
 * Create a mocking TelegrafContext object for unit testing
 * scene.
 *
 * @param {Object} update
 *   The mocked update object. Supposed to be sent by server.
 * @param {*} contextExtra
 *   Any extra attributes to the resulting context.
 * @returns {TelegrafContext}
 *   The mocked telegraf context for unit testing.
 */
function makeMockContext(update = {}, contextExtra = {}) {
  const tg = new Telegram('', {});
  tg.callApi = (method, data) => {
    console.log(`mocked tg callApi ${method}`, data);
  };

  const ctx = new Context(update, tg, {});
  Object.assign(
    ctx,
    {
      session: {},
      debug: {
        currentScene: '',
        replies: [],
        replyMessages: () => ctx.debug.replies.map(({ message }) => message),
      },
    },
    contextExtra,
  );
  ctx.reply = (message, extra = undefined) => {
    ctx.debug.replies.push({ message, extra });
  };
  ctx.scene = {
    enter: (sceneName) => {
      ctx.debug.currentScene = sceneName;
    },
    leave: () => {
      ctx.debug.currentScene = '';
    },
  };

  return ctx;
}

module.exports = {
  makeMockContext,
};

```

### Usage

In a testing framework like [jest](https://jestjs.io/), you can mock a context with a mock update object to test entering, normal messaging and leaving the scene:

```node
describe('myScene', () => {
  test('enter', async () => {
    const ctx = makeMockContext({ message: { chat: { id: 1234 } } });
    await chatScene.enterMiddleware()(ctx);

    // check the resulting reply array
    expect(ctx.debug.replyMessages()).toStrictEqual(['The message myScene sent on enter']);

    // Check the extra attributes with reply
    expect(ctx.debug.replies[0].extra).toStrictEqual({ somethingExtraOnEnter: true });
  });

  test('message', async () => {
    const ctx = makeMockContext({ message: { text: 'hello world', chat: { id: 1234 } } });
    await chatScene.middleware()(ctx);

    // check the resulting reply array
    expect(ctx.debug.replyMessages()).toStrictEqual(['myScene got your hello world!!!']);

    // Check the extra attributes with reply
    expect(ctx.debug.replies[0].extra).toStrictEqual({ somethingExtraOnMessage: true });
  });

  test('leave', async () => {
    const ctx = makeMockContext({ message: { text: 'bye bye', chat: { id: 1234 } } });
    await chatScene.leaveMiddleware()(ctx);

    // check the resulting reply array
    expect(ctx.debug.replyMessages()).toStrictEqual(['myScene got your hello world!!!']);

    // Check the extra attributes with reply
    expect(ctx.debug.replies[0].extra).toStrictEqual({ somethingExtraOnLeave: true });
  });
});
```
